serif_swing/src/main/kotlin/serif_swing/App.kt

```diff
@@ -208,13 +266,96 @@ class SwingChatRoom(val transition: (MatrixState, Boolean) -> Unit, val panel: J
                     } else {
                         JLabel(og_image_icon)
                     }
-                } else {
-                    val message = JTextArea(message)
+                }
+                is SharedUiAudioMessage -> {
+                    val audio_url = msg.url
+                    val play_btn = JButton("Play/Pause $audio_url")
+                    play_btn.addActionListener({
+                        AudioPlayer.loadAudio(audio_url)
+                        AudioPlayer.play()
+                    })
+                    play_btn
+                }
+                else -> {
+                    //val message = CustomTextPane(true)
+                    val message = JTextPane()
+                    message.setEditorKit(WrapEditorKit);
                     message.setEditable(false)
-                    message.lineWrap = true
-                    message.wrapStyleWord = true
+                    // This is mandatory to make it wrap, for some reason
+                    // It's not in the examples I found online
+                    // My best guess is that because of the layout, it
+                    // won't smash it smaller than preferred size, but it will
+                    // stretch it to fit the larger size?
+                    message.setPreferredSize(Dimension(0, 0))
+
+                    var current_idx = 0
+                    val simpleAttrs = SimpleAttributeSet()
+                    for (url_match in URL_REGEX.findAll(msg.message)) {
+                        println("${url_match.value} is a url!!")
```

Comment:
We should remove this print given the logging that occurs when attempting to open the link

Reply:
Fair, done!